core/mount/mount_linux.go

```diff
@@ -260,23 +263,23 @@ func doPrepareIDMappedOverlay(lowerDirs []string, usernsFd int) (tmpLowerDirs []
 			// Using os.Remove() so if it's not empty, we don't delete files in the
 			// rootfs.
 			if err := os.Remove(lowerDir); err != nil {
-				log.L.WithError(err).Warnf("failed to remove temporary overlay lowerdir's")
```

Comment:
Could you please remove the `unix.MNT_DETACH` at the line 256? I think it's not required if we use read-only.

Reply:
@fuweid sadly it is needed. There is some "contention" issue (the umount fails with busy when creating a lot of pods), see here for a repro: https://github.com/containerd/containerd/issues/10704#issuecomment-2372396341

I'm unsure yet why the layers seem to be re-used by other mounts or what. We should debug that and fix it, but until then, the detached mount is the best way around it.

Also, I'm not sure I'd remove it even after that is fixed. A detached mount just makes it more robust: if it is used for some reason, all following access will be disconnected and the unmount will be performed as soon as it is not being used anymore.

Reply:
I left it in for the reasons @rata stated. That said, I think @fuweid may understand the possibility of leaking mounts if we remove it and feels this is preferable to using MNT_DETACH, [see their prior comment about it](https://github.com/containerd/containerd/pull/10721#discussion_r1797425890).

> NOTE: Even if there are leaky mountpoint, containerd will clean it up during booting. 

I'm curious if there's a specific risk or downside to using MNT_DETACH that you are aware of @fuweid?

Reply:
Hi @mbaynton Sorry for late reply. I can't reproduce this issue by the script you provided. The EBUSY might be caused by other running processes, for example, security application scans all the files. When one process holds fd on that mountpoint, it's unlikely to umount it without `MNT_DETACH`.

Personally, I don't like MNT_DETACH because it hides real issue. We, contained, have similar issue like https://github.com/containerd/containerd/issues/5538. That's why I want to use `readonly` here.

Since we have cleanup during starting containerd, if it's caused by security application and it's transient issue, it won't be issue.

Reply:
Additionally, `MNT_DETACH` will hide mountpoint and we can't see it. If there are applications holding files in that mountpoint for a long time, we can say it's leaky issue. The application should be aware of this, instead of using `MNT_DETACH`.

Reply:
I've removed `MNT_DETACH` as you requested.

I agree that without knowing the real issue in detail (is it in fact because both my clusters and @rata's were running other software that jumped right on the new mounts, or is there another explanation) it is a bit speculation as to whether MNT_DETACH is doing what we think.

I would like to explore achieving these idmap mounts using the new mount API and overlayfs API extension to specify lower layers by fd, because I think this method will lead to greater isolation to just containerd and the kernel. Lower layer idmap mounts need not ever be attached to any mount namespace / directory tree. However, if that method does not work out, then I think it will be worth the tracing necessary to understand what is usually causing the EBUSY we originally saw. In that case, I _may_ come back and advocate for `MNT_DETACH` again :)

> I can't reproduce this issue by the script you provided.

Interesting, thanks for trying it out. You were using a build prior to #10721? This is good feedback, I only knew that it worked on a few of my clusters and it worked for @rata, I was starting to think it was a highly reliable repro.

> When one process holds fd on that mountpoint, it's unlikely to umount it without MNT_DETACH

I think with `MNT_DETACH` we have a chance for the mount to become unused and freed, where as with it, we ensure the mount remains indefinitely or at least until the restart cleanup you mentioned.

Reply:
> I would like to explore achieving these idmap mounts using the new mount API and overlayfs API extension to specify lower layers by fd (later, in other issues/PRs), because I think this method will lead to greater isolation to just containerd and the kernel.

+1 looking forward to that!

> Interesting, thanks for trying it out. You were using a build prior to https://github.com/containerd/containerd/pull/10721? This is good feedback, I only knew that it worked on a few of my clusters and it worked for @rata, I was starting to think it was a highly reliable repro.

Yes, I reverts #10721 in my local and I didn't reproduce this issue.